Feature collisions

Diocruel opened the pull request:
Finally a basic model of collisions is implemented in the framework!

Current adding this to the def branch will result in a test Coverage of 77%. The main problem at the moment is VRPlayer class in the model package. This class contains a few methods that could be tested with the help of intensive mocking, however these methods will change slightly when a more advanced camera model is introduced. The current model is too simplistic, but will most probably improve as soon as real VR support is implemented. 

Due to time constraints it will not be possible to add VR support this sprint( it was also not in the backlog of this sprint), so I suggest adding this pull request in order to show off the world Bram has created in FP.

Current Physics work by maintaining a BulletStateApp object which holds all physical objects in the game world. This class takes care of all collisions and also tests for impossible movements. The camera is still the flyby camera and is updated to be at the position of the player each update. 

~

Diocruel committed:
Created temporary class for jMonkey research

Used to follow some tutorials on jMonkey

Diocruel committed:
Changed name of collision class

Diocruel committed:
Added test Town scene.

Scene is used for manually testing code, should be removed
later.

Diocruel committed:
Copied most of the collision tutorial

Diocruel committed:
Fixed a few checkstyle errors

Can't find jBullet class

Geweldig committed:
Add JSON overhead

Creates an easy interface for loading/saving json files.

Geweldig committed:
Add JSONUtil tests

Creates tests for the JSONUtil.

Geweldig committed:
Modify JSONUtil test

This adds a test for reading of an existing file.
This bumps the coverage to 90%, the only line not tested
is the private constructor.

Geweldig committed:
Fix checkstyle

This fixes all checkstyle issues in JSONUtil and JSONUtilTest.

Taeir committed:
Remove codecov settings + fix travis mistake

Codecov settings file was misconfigured.
Travis must only push to codecov on successful builds.

ChielBruin committed:
Fixed Checkstyle warning.

There was an unused import.

ChielBruin committed:
Introduced controllers.

The main class now only manages a controller.
This controller is what runs the actual game.

ChielBruin committed:
Added pause option.

The main game can now be paused and resumed.

ChielBruin committed:
Fixed duplicate calls.

Initialize and cleanup were both called multpile times when switching
controllers.

ChielBruin committed:
Added --debugHud run argument

Added the option --debugHud to enable the statsView and fps counter.
These are now disabled by default.

ChielBruin committed:
Added inputListeners to the Controllers

Added methods to add input listeners to the controller classes.
Was already possible by calling Main.getInstance().getInputManager().[add/remove]Listener(),

ChielBruin committed:
Fixed issue with controller switching

The GameController did not call its superclass constructor and therefore
had an detached rootNode and guiNode.

ChielBruin committed:
Added Paused text

Added the text Paused to the Pause screen.

Diocruel committed:
Added libraries used in jBullet

Dependencies of jBullet were not included, so they are added in this
pull-request.

Diocruel committed:
Added a few spaces

ChielBruin committed:
Moved and fixed some tests

Diocruel committed:
Fixed all checkstyle errors

Diocruel committed:
Added collidable interface

Base interface structure.

Diocruel committed:
Added getSpatialWithPhysicsControl

Diocruel committed:
Clean up of java doc and comments

ChielBruin committed:
Merge branch 'dev-sprint3' into feature-collisions

ChielBruin committed:
Merge branch 'dev-sprint3' into feature-collisions

Diocruel committed:
Let entity implement Collidable

We want each entity to be collidable in our world so it does not fall
through the world.

Diocruel committed:
Implemented getSpatial in VRplayer.

Added final constants for physical interaction.

Diocruel committed:
Added some more final constants to VRplayer

Diocruel committed:
Added some test to VRPlayer to test spatial

Diocruel committed:
Added some more test to VRPlayerTest

Diocruel committed:
Implemented Collisions in controller

Due to earlier implementation drawable now extends Collidable.

Diocruel committed:
Added unimplemented method in MazeTile

Now is collidable.

Diocruel committed:
Some test changes, not this test fails.

Diocruel committed:
Merge remote-tracking branch 'origin/refactor-framework' into feature-collisions

Conflicts:
	src/main/java/nl/tudelft/contextproject/controller/Controller.java
	src/main/java/nl/tudelft/contextproject/model/VRPlayer.java

Diocruel committed:
Fixed failing tests

Diocruel committed:
Merge remote-tracking branch 'origin/dev-sprint3' into feature-collisions

Conflicts:
	src/main/java/nl/tudelft/contextproject/Main.java
	src/main/java/nl/tudelft/contextproject/controller/Controller.java
	src/test/java/nl/tudelft/contextproject/controller/ControllerTest.java
	src/test/java/nl/tudelft/contextproject/controller/GameControllerTest.java

Diocruel committed:
Fixed a lot of collision implementations

Diocruel committed:
More progress on implementing collisions

Diocruel committed:
More progress on collisions

Diocruel committed:
Collisions now work, but camera turning doesnt

Progress!

Diocruel committed:
Fixed camera movement finally

Diocruel committed:
Update spatial location based on movement of cam

Diocruel committed:
Fixed failing test

Note that some classes can't be tested untill VR support is included.

Diocruel committed:
Merge remote-tracking branch 'origin/dev-sprint3' into feature-collisions

Conflicts:
	src/main/java/nl/tudelft/contextproject/Main.java
	src/main/java/nl/tudelft/contextproject/model/level/MazeTile.java
	src/test/java/nl/tudelft/contextproject/model/level/MazeTileTest.java

Diocruel committed:
Fixed failing tests due to new Axis allignment

Diocruel committed:
Removed not needed line of code.

Diocruel committed:
Fixed quite a few issues with walls

Diocruel committed:
Fixed failing test, 23rd time.

Diocruel committed:
Changed some gravity parameters for Demo

Diocruel committed:
Removed Tutorial class

Diocruel commented:
I'll fix the pull request after travis is done...
EDIT: Done.

Diocruel committed:
Merge remote-tracking branch 'origin/dev-sprint3' into feature-collisions

Conflicts:
	src/main/java/nl/tudelft/contextproject/Main.java

Taeir commented on pom.xml:
```diff
@@ -146,6 +146,35 @@
+			<scope>system</scope>
+			<systemPath>${basedir}/jme_vr_libs/jbullet.jar</systemPath>
+		</dependency>
 	</dependencies>
```
I think master already has this done in a different way, that will be a nice merge conflict. (But this is the correct code for this PR).

Diocruel replied:
Did you add these libraries after we discussed them to the master then? Or did you change the way maven loads the libraries?

Taeir replied:
The latter. Master already includes these libraries, and loads them in a different (better) way.
The comment is more useful for the person merging this into master. The pom.xml of master is correct and does not need to be updated.

Taeir commented on src/main/java/nl/tudelft/contextproject/Main.java:
```diff
@@ -118,11 +119,16 @@ protected void setInputManager(InputManager im) {
-		getFlyByCamera().setMoveSpeed(50);
+		
+		flyCam.setMoveSpeed(100);
+		viewPort.setBackgroundColor(new ColorRGBA(0.1f, 0.1f, 0.1f, 1f));
```
Why not use getFlyByCamera? The getters can be properly mocked, the field cannot.
In addition, jMonkeyVR does some strange stuff with cameras, and explicitly mentions that you should use the getters for proper VR support.

Diocruel replied:
I can change it if you'd prefer it. All tutorials and examples used the field. This wil also be removed when VR support is added as this won't use a flyby camera.

Geweldig replied:
If you are certain the VR camera won't use this I'm fine with it staying like this.

Taeir commented on src/main/java/nl/tudelft/contextproject/Main.java:
```diff
@@ -133,6 +139,20 @@ public void simpleInitApp() {
+		inputManager.addMapping("Up", new KeyTrigger(KeyInput.KEY_W));
+		inputManager.addMapping("Down", new KeyTrigger(KeyInput.KEY_S));
+		inputManager.addMapping("Jump", new KeyTrigger(KeyInput.KEY_SPACE));
+	}
```
Should use getInputManager() for mocking and testing reasons. (For example, TestUtil properly instantiates getInputManager(), so this method could be tested by it).

BCrlrd replied:
Agreed

Taeir commented on src/main/java/nl/tudelft/contextproject/model/Collidable.java:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * Collidable interface.
+ *	Each object that uses some kind of physics should implement this class.
```
Indentation

Geweldig replied:
class = interface

Taeir commented on src/main/java/nl/tudelft/contextproject/model/Drawable.java:
```diff
@@ -8,7 +8,7 @@
  * This object has a geometry that can be updated.
  */
-public interface Drawable {	
+public interface Drawable extends Collidable {	
```
This is a potential code smell, as it implies that every object that we could possibly draw on the screen is solid (non pass through).

Diocruel replied:
@chiel prefered it this way. A collidable does not mean that the entire object is solid. I purpose a name change of the interface to physicalinteraction

Taeir replied:
Hmm, I disagree with chiel then. It stands to reason that every object we draw to the screen COULD be collidable, but that does not mean that every object IS collidable.
This method removes the possibility of using `if (something instanceof Collidable)` to check if something collides with it.

We should rename either Drawable or Collidable, like @Diocruel suggests.

Taeir commented on src/main/java/nl/tudelft/contextproject/model/VRPlayer.java:
```diff
@@ -14,29 +17,65 @@
+	public static final float PLAYER_STEP_HEIGHT = 0.1f;
+	public static final float PLAYER_RADIUS = .5f;
+	public static final float PLAYER_HEIGHT = 3f;
+	public static final int PLAYER_AXIS = 1;
```
Nice use of constants!

Taeir commented on src/main/java/nl/tudelft/contextproject/model/VRPlayer.java:
```diff
@@ -14,29 +17,65 @@
+		Vector3f camLeft = Main.getInstance().getCamera().getLeft();
+		walkDirection = new Vector3f();
+		if (left) {
+			walkDirection.addLocal(camLeft.mult(.0235f));
```
`.0235f` could be a constant.

Taeir replied:
Shouldn't this one also call normalizeLocal()?

BCrlrd replied:
Agreed with the constant comment.

Taeir commented on src/main/java/nl/tudelft/contextproject/model/VRPlayer.java:
```diff
@@ -54,4 +93,62 @@ public void mapDraw(Graphics2D g, int resolution) {
+	 */
+	@Override
+	public void onAction(String name, boolean isPressed, float tpf) {
+		if (name.equals("Left")) {
```
Could potentially be a string switch.
(If people agree, in eclipse this can be changed automatically: Select the entire if + else ifs, press ctrl + 1, choose convert to switch)

Geweldig replied:
Either changing to a switch with a default case or adding an else would be great.

Taeir commented on src/main/java/nl/tudelft/contextproject/model/level/MazeTile.java:
```diff
@@ -30,63 +35,85 @@
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid TileType: " + type);
 		}
```
I would opt that the original indentation is more correct. I do not know why eclipse by default formats the code without indentation increase for switch.

(However, this is very minor and possibly a personal habit/preference. Either way is fine by me)

BCrlrd replied:
as I wrote the original indentation, I agree :p

Taeir commented on src/main/java/nl/tudelft/contextproject/model/level/MazeTile.java:
```diff
@@ -30,63 +35,85 @@
-        this.spatial.move(position.x, position.y, height);
-        return spatial;
+		this.spatial.setMaterial(mat);                   // set the cube's material
+		this.spatial.move(position.x, height, position.y);
```
Well caught, this small parameter mess up could have caused a ton more problems...

Geweldig replied:
not so much a mess up as a change of implementation

Taeir replied:
Ah I see

Diocruel replied:
gravity works in strange ways ╭(۝乁 ͜ﾚ ㄏ۝)╮

Taeir commented on src/test/java/nl/tudelft/contextproject/controller/ControllerTest.java:
```diff
@@ -104,7 +113,11 @@ public Spatial getSpatial() {
+				return null;
+			}
+			
+			public void mapDraw(Graphics2D g, int resolution) { }
```
No `@Override` above mapDraw any more?

Taeir commented on src/test/java/nl/tudelft/contextproject/controller/ControllerTest.java:
```diff
@@ -129,7 +142,11 @@ public Geometry getSpatial() {
+				return null;
+			}
+			
+			public void mapDraw(Graphics2D g, int resolution) { }
```
No `@Override` above mapDraw any more?

Taeir commented on src/test/java/nl/tudelft/contextproject/controller/ControllerTest.java:
```diff
@@ -155,7 +172,11 @@ public Spatial getSpatial() {
+				return null;
+			}
+			
+			public void mapDraw(Graphics2D g, int resolution) { }
```
No `@Override` above mapDraw any more?

Diocruel replied:
Did you just comment 3 times on the same issue? Or is it 3 times more important then the rest of the issues?

Geweldig replied:
@Taeir really wanted to drive the point home

Taeir commented on src/test/java/nl/tudelft/contextproject/controller/GameControllerTest.java:
```diff
@@ -57,26 +60,34 @@ public void setUp() {
+
 		controller.setGame(game);
+		controller.setPhysicsEnvironmentNode(phe);
 	}
```
Ouch, this setup method is getting pretty extensive. This could possibly be cleaned up when we have more time.

Geweldig commented on src/main/java/nl/tudelft/contextproject/model/VRPlayer.java:
```diff
@@ -54,4 +93,62 @@ public void mapDraw(Graphics2D g, int resolution) {
+	 * On button press, what the player should do.
+	 * @param name
+	 * @param isPressed
+	 * @param tpf
```
(nitpicking) Explaining what the parameters stand for.

Taeir replied:
This should also cause a checkstyle warning, so perhaps not so nitpicking

Diocruel replied:
I will add the constants doc at the top of the page, I did not receive a check style warning here.

Taeir commented:
Well, quite a pull request do we have here.

There are a few small things about styling that I created line comments about.
There are also some potential code smells (that should not be fixed in this PR). It would be useful to add TODO tags there so our IDE's remind us that there is still work to do there.

As a final note, the automatic formatting and indentation fixing that was applied fixes styling, yes, but it also causes a ton of additional changes that are not necessary and not relevant to this PR.
In addition, all these small changes have a high likelihood of creating merge conflicts.

So, for next time I would advise to try and limit usage of CTRL+I and CTRL+SHIFT+F to branches created specifically for improving code layout/styling.

When the issues I have mentioned have been addressed, I'd say that this can be merged.

(Poor soul that's going to merge this into master tomorrow).

Diocruel commented:
I won't try to change other people code anymore then, especially with CRTL + I. I only did it because I thought it was required that all code would adhere to the same style. Fixing stuff for the next 1.5 hours.

Diocruel committed:
Updated main with PR requests

Diocruel committed:
Renamed Collidable to PhysicsObject

Diocruel committed:
Extropelated some constants

Diocruel committed:
Final fixes to the PR

Diocruel committed:
Taico change

Diocruel committed:
Removed unnecessary javadoc

Diocruel commented:
Fixed the following stuff:
Main: used getters instead of fields.
Collidable: renamed interface, fixed javadoc.
VRPlayer: Extrapolated constants, explained Constants, removed javadoc, implemented switch instead of if else tree. 
Will add TODOs now

Diocruel committed:
Added TODOs for changes when VR is implemented

Geweldig commented:
Looks gooed, should be merged!

Taeir commented:
This can be merged now.
